core/src/consensus.rs

```diff
@@ -1042,46 +1043,86 @@ impl Tower {
         self.last_switch_threshold_check.is_none()
     }
 
-    /// Performs threshold check for `slot`
-    ///
-    /// If it passes the check returns None, otherwise returns Some(fork_stake)
-    pub fn check_vote_stake_threshold(
+    /// Checks a single vote threshold for `slot`
+    fn check_vote_stake_threshold(
+        vote: Option<&Lockout>,
+        vote_state: &VoteState,
+        threshold_depth: usize,
+        threshold_size: f64,
+        slot: Slot,
+        voted_stakes: &HashMap<Slot, u64>,
+        total_stake: u64,
+    ) -> ThresholdDecision {
+        let vote = match vote {
+            None => return ThresholdDecision::PassedThreshold,
+            Some(vote) => vote,
+        };
+        let fork_stake = match voted_stakes.get(&vote.slot()) {
+            None => {
+                // We haven't seen any votes on this fork yet, so no stake
+                return ThresholdDecision::FailedThreshold(threshold_depth as u64, 0);
+            }
+            Some(fork_stake) => fork_stake,
+        };
+
+        let lockout = *fork_stake as f64 / total_stake as f64;
+        trace!(
+            "fork_stake slot: {}, vote slot: {}, lockout: {} fork_stake: {} total_stake: {}",
+            slot,
+            vote.slot(),
+            lockout,
+            fork_stake,
+            total_stake
+        );
+        if vote.confirmation_count() as usize > threshold_depth {
+            for old_vote in &vote_state.votes {
+                if old_vote.slot() == vote.slot()
+                    && old_vote.confirmation_count() == vote.confirmation_count()
```

Comment:
@carllin  do you remember why we do this? I found https://github.com/solana-labs/solana/pull/7948 but since `voted_stakes` is rolled up across forks it shouldn't matter right?

Reply:
I believe this was the issue: https://github.com/solana-labs/solana/issues/7904#issuecomment-577441985

If you bounce back to voting on the main fork after not voting for a while, your latest vote `N` on the main fork will pop off a lot of the stake of votes in the tower that would have rolled up to earlier votes in the tower. I think this is because for every vote state we apply the proposed vote `N` latest before roling up to `voted_stakes`